### ISSUE TITLE
no relative paths in subdocuments

### ISSUE THREAD

**Issue by getreu** [asciidoctor/asciidoctor#650]
I am using the _include::_ directive to divide larger documents in independent subdocuments. Each subdocument has it's own resources and own image folders.   
1. First all subdocuments are processed individually.
2. Once all subdocuments are complete they are assembled by a master document using _include::_  and   _:leveloffset: 1_ 

Unfortunately image paths are relative to _docdir_. I step 1 _docdir_ points to the directory of the subdocument. In step 2 it points to the directory of the master document. Hence in step 2 the images are not found any more. 

Is there a way to set the _images_ attribute to the directory of the current subdocument?  In asciidoc exists _indir_ which is the directory of the current document but I did not find anything similar in asciidoctor.

Thanks in advance.

**Comment by mojavelinux** [comment]
The problem with that approach is that includes are not subdocuments. They are preprocessor macros. Therefore there is no place to add any extra information in the AST. (We have no idea where they will be placed and it may not be in a place where information can be stored, such as in the middle of a verbatim block).

If you want subdocuments, then you'd need to develop a custom block macro using the extension API. A block macro is significant because it represents a block AST node, where there is an insertion point to add additional context. That is possible today, though we may discover limitations that need addressing in the process.

**Comment by elextr** [comment]
I am not an expert on the Asciidoctor implementation details, but IIUC the problem is that the `image::` directive is not processed until well after the include is processed.  So the original included file information may no longer be available.  It would have to be recorded at inclusion time, which is probably an extension to existing Asciidoctor handling although I did suggest [here](https://github.com/asciidoctor/asciidoctor/issues/894#issuecomment-430033423) that the information may be extractable but don't know if thats feasible.

**Comment by johannesjh** [comment]
Ok, thank you @mojavelinux  for pointing it out; this will indeed make it difficult to implement the functionality as an extension. 

An alternative solution is to change how the `include` preprocessor macro is processed, in order to set and reset variables, similar to how `leveloffset` is implemented. I implemented the following proof-of-concept:

```diff
diff --git a/lib/asciidoctor/reader.rb b/lib/asciidoctor/reader.rb
index cd37cb94..8d32c655 100644
--- a/lib/asciidoctor/reader.rb
+++ b/lib/asciidoctor/reader.rb
@@ -1146,28 +1146,35 @@ class PreprocessorReader < Reader
       # FIXME we eventually want to handle leveloffset without affecting the lines
       if attributes.key? 'leveloffset'
         @lines.unshift ''
         @lines.unshift %(:leveloffset: #{attributes['leveloffset']})
         @lines << ''
         if (old_leveloffset = @document.attr 'leveloffset')
           @lines << %(:leveloffset: #{old_leveloffset})
         else
           @lines << ':leveloffset!:'
         end
         # compensate for these extra lines
         @lineno -= 2
       end
 
+      # Expose the included file's name and dir as variables
+      @lines.unshift ''
+      @lines.unshift %(:current_docfile: #{@include_stack&.dig(-1, 1)})
+      @lines.unshift %(:current_docdir: #{@include_stack&.dig(-2, 2)})
+      @lines << %(:current_docfile: #{@include_stack&.dig(-2, 1)})
+      @lines << %(:current_docdir: #{@include_stack&.dig(-2, 2)})
+
       # FIXME kind of a hack
       #Document::AttributeEntry.new('infile', @file).save_to_next_block @document
       #Document::AttributeEntry.new('indir', @dir).save_to_next_block @document
       @look_ahead = 0
     end
     self
   end
 
   def create_include_cursor file, path, lineno
     if ::String === file
       dir = ::File.dirname file
     elsif ::RUBY_ENGINE_OPAL
       dir = ::File.dirname(file = file.to_s)
     else
```

Would a solution like this be acceptable as pull request?

**Comment by mojavelinux** [comment]
> An alternative solution is to change how the include preprocessor macro is processed

The include processor is an extension point, so you can accomplish what you're proposing using an extension. (See https://asciidoctor.org/docs/user-manual/#include-processor-example).

> Would a solution like this be acceptable as pull request?

This doesn't belong in core because it has side effects that would affect existing documents. That's why I deferred implementing #2608.

The reason we can make an exception for leveloffset is because we're assuming that if you specify leveloffset, you must be using the include directive at a section boundary. In other words, there's a place to inject an attribute entry. That assumption is not universal.

As I've stated in other issues related to this one, if we want subdocuments in AsciiDoc, then we need to implement a subdocument macro. An include directive can never be a subdocument because it gets expanded before the document is parsed and thus has no awareness of where it is located in the document hierarchy. (But you *could* assume that to be the case if you make your own include processor extension).

**Comment by mojavelinux** [comment]
> In order to solve this issue quickly without the need for long discussions about backwards compatibility

IMO, that's not the right way to solve problems in software. We need to architect or else we are just throwing hacks around. (But hacks are perfect for extensions, which is why extensions exist).

**Comment by mojavelinux** [comment]
Btw, in Asciidoctor 1.5.8, the imagesdir attribute is now stored on the image node itself. So an extension can now look at the target and imagesdir attribute on the node and mutate the target accordingly. (The next step would be to have the image_uri helper look at the imagesdir attribute on the node first, which would allow you to influence the resolved path per node).

**Comment by johannesjh** [comment]
Thank you very much for these clarifications. I agree that a proper solution for a subdocument macro would be much better than a quick hack. I just hesitated about proposing or creating a proper solution because I guessed it will take a long time to define and implement, given the amount discussion that has already gone into it. 

I have, for now, aimed at creating a quick solution and implemented this extension:
https://github.com/johannesjh/asciidoctor-extensions-lab/blob/indir/lib/indir-include-processor/extension.rb, allowing me to write:

```adoc
= Sub Document

This document can be compiled as standalone document, or included into another document.

ifndef::indir[:indir: .]
image::{indir}/images/example.svg[]
```

I would like to hear your feedback regarding the extension. I am willing to change and improve things and submit a pull request.

**Comment by mojavelinux** [comment]
The extension you wrote is consistent with what I had in mind. Feel free to submit that to the extension lab for reference because I think it will help people looking for a starting point.

You may also be interested in leveraging https://github.com/jirutka/asciidoctor-include-ext, which externalizes the built-in logic for handling an include so that you don't have to reimplement everything when writing an include processor. I plan to make that logic reusable, but until then, that library could help.

**Comment by johannesjh** [comment]
I sent pull request https://github.com/asciidoctor/asciidoctor-extensions-lab/pull/100 with the extension plus some documentation.

Thank you for pointing me at https://github.com/jirutka/asciidoctor-include-ext - this extension looks very nice, it allows for finer-grained extensions of asciidoctor's include behavior.

**Comment by johannesjh** [comment]
I tried using https://github.com/jirutka/asciidoctor-include-ext for the extension, but it did not really make the code cleaner.

Straightforward implementation without [asciidoctor-include-ext](https://github.com/jirutka/asciidoctor-include-ext):
```ruby
    # this solution inherits directly from asciidoctor
    class IndirIncludeProcessor < Asciidoctor::Extensions::IncludeProcessor
      def process doc, reader, target, attributes
        content = (open target).readlines

        # Set variables at beginning of the included content
        included_docfile = target
        included_docdir = ::File.dirname target
        content.unshift ''
        content.unshift %(:indir: #{included_docdir})

        # Reset the variables at the end of the included content
        parent_docfile = doc.reader.include_stack&.dig(-1, 1) || (doc.attr 'docfile')
        parent_docdir = ::File.dirname parent_docfile
        content << ''
        content << %(:indir: #{parent_docdir})

        # Push included content for further processing
        reader.push_include content, target, target, 1, attributes
        reader
      end
    end
```

Alternative implementation based on [asciidoctor-include-ext](https://github.com/jirutka/asciidoctor-include-ext):

```ruby
    # this solution inherits from the asciidoctor-include-ext extension
    class IndirIncludeProcessor2 < Asciidoctor::IncludeExt::IncludeProcessor
      def initialize(*args, &block)
        # temporary storage helper that won't be frozen by Asciidoctor
        @tmp = { }
        super
      end

      def process(document, reader, target, attributes)
        @tmp[:document] = document
        @tmp[:reader] = reader
        @tmp[:target] = target
        super
      end

      def read_lines(filename, selector)
        # read content using functionality from super
        content = super(filename, selector)
        if content.is_a? String then content = content.lines end

        # Set variables at the beginning of the included content
        included_docfile = @tmp[:target]
        included_docdir = ::File.dirname @tmp[:target]
        content.unshift ''
        content.unshift %(:indir: #{included_docdir})

        # Reset the variables at the end of the included content
        parent_docfile = @tmp[:reader].include_stack&.dig(-1, 1) || (@tmp[:document].attr 'docfile')
        parent_docdir = ::File.dirname parent_docfile
        content << ''
        content << %(:indir: #{parent_docdir})
        return content
      end
    end
```

The above code shows the problem in using [asciidoctor-include-ext](https://github.com/jirutka/asciidoctor-include-ext) for my purpose. The problem is that asciidoctor-include-ext does not pass enough information to the `read_lines` method that I need to overwrite. As a workaround, I tried passing the information around in member variables, but this did not work because Asciidoctor freezes the object. Hence the ugly workaround with a `@tmp` object, the contents of which do not get frozen. This has potential for dangerous race conditions if a single instance of the extension is kept for compiling multiple documents. 

In conclusion, I don't think I will invest more time into an extension based on [asciidoctor-include-ext](https://github.com/jirutka/asciidoctor-include-ext). Extending an extension is just too fragile - it is nicer to work against asciidoctor core. Also, instead of investing more time into building an extension, it would be more worthwile to think about a proper subdocument directive.

**Comment by johannesjh** [comment]
Note: I published the extension as a ruby gem
https://github.com/johannesjh/asciidoctor-indir_ext

**Comment by Hollerweger** [comment]
> Note: I published the extension as a ruby gem https://github.com/johannesjh/asciidoctor-indir_ext

Thanks a lot for the extension. Could it me integrated in Asciidoctor itself?
It should be supported out of the box with previews on Git/Gitlab or with standard plugin configuration on VSCode or IntellJ.
Or is there any alternative/workaround that is now supported?

**Comment by Hollerweger** [comment]
For images following condition can be used in the sub files to override the image dir:
ifndef::imagesdir[:imagesdir: ../images]

Could't find a similar approach for other references like plantuml diagrams.

**Comment by johannesjh** [comment]
+1 for "could this be integrated into asciidoctor itself?" ...the handling of relative paths in included documents really isn't something that should be done by a plugin.

**Comment by mojavelinux** [comment]
I have already answered this question:

> This doesn't belong in core because it has side effects that would affect existing documents. That's why I deferred implementing #2608.

**Comment by johannesjh** [comment]
That's right, deferred because it would be a major change with side effects. In #2608, you directed further discussions towards #3603, explaining that a new document-include-directive (which handles relative paths) should first be clarified with the upstream asciidoc language, as opposed to asciidoctor's language implementation. I agree. Clarifying this at the language level would have the benefit of interoperability between all language implementations.

What to do next? - I propose to close this issue in favor of #3603. To move the issue forward, I guess someone would have to go clarify a document-include-directive with the upstream asciidoc language. We can discuss this in #3603.

**Comment by mojavelinux** [comment]
I don't see a reason to close this issue at this time as we're currently waiting on other initiatives to evolve, specifically the AsciiDoc Language project (which won't happen overnight and maybe even take years). The state of this issue reminds me that this is a topic which needs to be addressed in some way. If I close it, I will never look at it again.
